### ISSUE TITLE
WebKit2?

### ISSUE THREAD

**Issue by quozl** [michaldaniel/Ebook-Viewer#34]
G'day Michal,

For your interest, I'm maintaining another viewer, the [Read](https://github.com/sugarlabs/read-activity) activity at [Sugar Labs](https://www.sugarlabs.org/), which is bundled on the [One Laptop per Child](http://one.laptop.org/) laptops.  Our viewer like yours is Python GTK+ and uses WebKit to render EPUB files, and also uses Evince to render PDF files.  See [User Manual](https://help.sugarlabs.org/en/read.html).

We're faced with the impending doom of obsolescence of the WebKit 3.0 API.  Are you planning a port to WebKit2 4.0 API?

**Comment by ormung** [comment]
I looked a bit into this, porting should be straightforward, except for saving the scroll position.. that's gonna be a [huge pain](https://stackoverflow.com/questions/21781868/scrolling-a-webkit2-webkit-window-in-gtk3)

**Comment by quozl** [comment]
Thanks.  Yes, having control of scroll position was a huge pain, requiring asynchronous JavaScript, and I gave up on it after a week or so, ended up removing EPUB support from Read activity in Sugar, sadly.

**Comment by fabiocolacio** [comment]
@quozl @ghost, I maintain a markdown editor called [Marker](https://github.com/fabiocolacio/Marker), which uses webkit2gtk-4.0 for a live HTML preview.

I too struggled a lot with controlling the scroll position using Javascript. I was having an issue where refreshing the web page would scroll to the top, and I wanted the page position to be saved. I eventually found a relatively simple solution, by writing a WebExtension.

[Here](https://github.com/fabiocolacio/Marker/blob/06c4e27275435c1d9e561b730546a3ff331a67c6/src/webkit-extension/scroll-extension.c) is one of the first working versions of that WebExtension for reference.

Maybe this can help you out?

**Comment by quozl** [comment]
Thanks @fabiocolacio!

**Comment by ArhantJain** [comment]
Is this issue open? If Yes, plz guide me, about solving.

**Comment by quozl** [comment]
@ArhantJain, you will see in https://github.com/michaldaniel/Ebook-Viewer/blob/52c9d0f530557882b254f38c2eac68e25d930155/src/components/viewer.py#L19 that WebKit 3.0 API is still used.  Suggested steps for solving;

- [ ] read and understand the WebKit 3.0 API using the upstream documentation, focusing on the classes and methods used in Ebook-Viewer source code,
- [ ] list the functionality of WebKit 3.0 API used by Ebook-Viewer source code,
- [ ] read and understand the WebKit2 4.0 API using the upstream documentation, focusing on the functionality provided by the library,
- [ ] find the commonality; a mapping from WebKit 3.0 API to WebKit2 4.0 API,
- [ ] edit the source code of Ebook-Viewer to use the WebKit2 4.0 API,
- [ ] iteratively test and repeat.

**Comment by JonasT** [comment]
For what it's worth, I just tried to install Ebook-Viewer on Fedora but I can't figure out how to do it since there's only webkitgtk4 as a package...

Is there any workaround for this?

**Comment by quozl** [comment]
Not really.  Fedora dropped WebKit 3.0 API.  You might rebuild the packages from an older release, but it will get harder and harder to do this as Fedora changes.
